### PR TITLE
Simplifications based on a super-struct for suite and spec

### PR DESCRIPTION
**Pull request by snogge**
Extracted a super struct buttercup-suite-or-spec from buttercup-suite and buttercup-spec, and then used it to simplify code and increase reuse.

### CODE REVIEW

File: buttercup.el [hunk 1]
```diff
@@ -730,26 +707,32 @@ See also `buttercup-define-matcher'."
...
   (buttercup-suites-total-specs-status suite-list 'failed))
 
+(defun buttercup--specs (spec-or-suite-list)
+  "Return a flat list of all specs in SPEC-OR-SUITE-LIST."
+  (let (specs)
+    (dolist (spec-or-suite spec-or-suite-list specs)
+      (if (buttercup-spec-p spec-or-suite)
+          (setq specs (append specs (list spec-or-suite)))
+        (setq specs (append specs (buttercup--specs
+                                   (buttercup-suite-children spec-or-suite))))))))
```
Comment (jorgenschaefer) [inline buttercup.el]: Not that it matters much performance-wise, but this is inefficient and highly unidiomatic – could you use the usual push/nreverse or (better, I think) the cl-loop pattern? :-)

File: buttercup.el [hunk 2]
```diff
@@ -730,26 +707,32 @@ See also `buttercup-define-matcher'."
...
 (defun buttercup-suite-full-name (suite)
   "Return the full name of SUITE, which includes the names of the parents."
-  (let ((name-parts (mapcar #'buttercup-suite-description
-                            (cons suite (buttercup-suite-parents suite)))))
-    (mapconcat #'identity
-               (reverse name-parts)
-               " ")))
+  (mapconcat #'buttercup-suite-description
+             (nreverse (cons suite (buttercup-suite-parents suite)))
+             " "))
```
Comment (jorgenschaefer) [inline buttercup.el]: Woah, nice catch!

File: buttercup.el (hunk 1 above)
Comment (snogge) [reply buttercup.el]: I think I tried both approaches and the problem is keeping the order.  buttercup--specs-and-suites is equally unappealing.

I'll have a think and see if I can come up with something.

Comment (jorgenschaefer) [reply buttercup.el]: Hm. Ah yes, tree recursive stuff. I think I'm actually fine with this then. Thank you!

**Comment by jorgenschaefer** [comment]
I'm happy to merge this now – anything you want to add or change in the branch still?

**Comment by snogge** [comment]
I have nothing more.

**Comment by jorgenschaefer** [comment]
Thank you!

**Pull request merged**